packages/utils/src/synergy.ts

```diff
@@ -328,92 +276,192 @@ export async function getSubgroupItems(perspective, subgroupId) {
+          newSubgroup.summary = newSubgroupSummary;
+          await newSubgroup.save();
+          // link subgroup to channel for use in search
+          await perspective.add({
+            source: channelId,
+            predicate: "ad4m://has_child",
+            target: newSubgroup.baseExpression,
+          });
+          // remove link to old subgroup if present
+          if (linkExpression) await perspective.remove(linkExpression);
+          // add link to new subgroup
+          await perspective.add({
+            source: newSubgroup.baseExpression,
+            predicate: "ad4m://has_child",
+            target: item.baseExpression,
+          });
+          subgroup = newSubgroup;
+        } else {
+          // if the subject of the conversation has stayed the same, stick with the exising subgroup
+          if (newSubgroupName) subgroup.subgroupName = newSubgroupName;
+          if (newSubgroupSummary) subgroup.summary = newSubgroupSummary;
+          if (newSubgroupName || newSubgroupSummary) await subgroup.update();
+        }
+      }
+      // attach topics to new item, conversation, & subgroup (avoid duplicates on conversation & subgroup)
+      const filteredTopics = topics.filter((topic: any) => topic && topic.name && topic.relevance);
+      await Promise.all(
+        filteredTopics.map(
+          (topic: any) =>
+            new Promise(async (resolve2: any) => {
+              // link new item topics
+              const topicId = await findOrCreateTopic(perspective, allTopics, topic.name);
+              await linkTopic(perspective, item.baseExpression, topicId, topic.relevance);
+              // link new conversation topics
+              const conversationTopics = await findTopics(perspective, conversation.baseExpression);
+              if (!conversationTopics.find((t) => t.name === topic.name)) {
+                await linkTopic(perspective, conversation.baseExpression, topicId, topic.relevance);
+              }
+              // link new subgroup topics
+              const subgroupTopics = await findTopics(perspective, subgroup.baseExpression);
+              if (!subgroupTopics.find((t) => t.name === topic.name)) {
+                await linkTopic(perspective, subgroup.baseExpression, topicId, topic.relevance);
+              }
+              resolve2();
+            })
+        )
+      );
+      // update embeddings
+      await removeEmbedding(perspective, item.baseExpression);
+      await createEmbedding(perspective, item.text, item.baseExpression);
+      if (!existingItem) {
+        await removeEmbedding(perspective, conversation.baseExpression);
+        await createEmbedding(perspective, newConversationSummary, conversation.baseExpression);
+        await removeEmbedding(perspective, subgroup.baseExpression);
+        await createEmbedding(perspective, newSubgroupSummary, subgroup.baseExpression);
+      }
 
-    resolve();
+      resolve();
+    }
```

Comment:
This is a pretty long function with several blocks with different concerns. It would help to extract meaningful functions...

Reply:
for instance having one function to collect all the needed data like the subgroup and previous messages..